### PR TITLE
[Compute] VirtualMachineScaleSetId Parameter for Update-Azvm and Reduced File permissions for SSH key

### PR DESCRIPTION
**Pull request by haagha**
## Description

Adding new parameter VirtualMachineScaleSetId and reducing permissions for private key when generating SSH key.

(https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/1374)

https://github.com/Azure/azure-powershell/issues/16843
## Checklist

- [ ] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

**Comment by grizzlytheodore** [comment]
also need changelog

File: src/Compute/Compute/VirtualMachine/Operation/UpdateAzureVMCommand.cs [hunk 1]
```diff
@@ -223,7 +228,9 @@ public override void ExecuteCmdlet()
                         Host = this.IsParameterBound(c => c.HostId)
                              ? new SubResource(this.HostId)
                              : this.VM.Host,
-                        VirtualMachineScaleSet = this.VM.VirtualMachineScaleSet,
+                        VirtualMachineScaleSet = this.IsParameterBound(c => c.VirtualMachineScaleSetId)
+                                                ? new SubResource(this.VirtualMachineScaleSetId)
```
Comment (grizzlytheodore) [inline src/Compute/Compute/VirtualMachine/Operation/UpdateAzureVMCommand.cs:232]: what happens if -VirtualMachineScaleSetId and -VM are both provided?  
Then here in this line -VM gets overwritten. 

It should be using -VM and only the vm.virtualmachinescaleset.id part gets overwritten, right?

Comment (haagha) [reply src/Compute/Compute/VirtualMachine/Operation/UpdateAzureVMCommand.cs:232]: The property VM.VirtualMachineScaleSet is  a subresource. So when user want's to attach to a new VMSS by providing VirtualmachinescalesetId parameter it should overwrite the VM.VirtualMachineScaleSet property.

Comment (grizzlytheodore) [reply src/Compute/Compute/VirtualMachine/Operation/UpdateAzureVMCommand.cs:232]: I see. the only property of subresource is ID, so makes sense.

File: src/Compute/Compute/VirtualMachine/Operation/UpdateAzureVMCommand.cs [hunk 2]
```diff
@@ -103,6 +103,11 @@ public class UpdateAzureVMCommand : VirtualMachineBaseCmdlet
         [AllowEmptyString]
         public string ProximityPlacementGroupId { get; set; }
 
+        [Parameter(
+            Mandatory = false)]
```
Comment (Sandido) [inline src/Compute/Compute/VirtualMachine/Operation/UpdateAzureVMCommand.cs:107]: Include a helpMEssage description.

File: src/Compute/Compute/VirtualMachine/Operation/UpdateAzureVMCommand.cs (hunk 1 above)
Comment (Sandido) [reply src/Compute/Compute/VirtualMachine/Operation/UpdateAzureVMCommand.cs:232]: that is confusing, but it is true. Just the naming is odd.

File: src/Compute/Compute/VirtualMachine/Operation/UpdateAzureVMCommand.cs [hunk 3]
```diff
@@ -244,6 +251,11 @@ public override void ExecuteCmdlet()
                         parameters.ProximityPlacementGroup.Id = null;
                     }
 
+                    if (parameters.VirtualMachineScaleSet != null && string.IsNullOrWhiteSpace(parameters.VirtualMachineScaleSet.Id))
```
Comment (Sandido) [inline src/Compute/Compute/VirtualMachine/Operation/UpdateAzureVMCommand.cs:254]: what does this do? Can you add a comment describing why you are setting a null or whitespace string to null?

File: src/Compute/Compute/help/Update-AzVM.md [hunk 4]
```diff
@@ -419,6 +421,21 @@ Accept pipeline input: True (ByPropertyName)
 Accept wildcard characters: False
 ```
 
+### -VirtualMachineScaleSetId
+Id for the Virtual Machine ScaleSet that the virtual machine should be updated to.
```
Comment (Sandido) [inline src/Compute/Compute/help/Update-AzVM.md:425]: Scale Set

**Review by Sandido — CHANGES_REQUESTED** [review]

File: src/Compute/Compute/VirtualMachine/Operation/UpdateAzureVMCommand.cs (hunk 1 above)
Comment (haagha) [reply src/Compute/Compute/VirtualMachine/Operation/UpdateAzureVMCommand.cs:232]: The naming is is that way because it's a subresource.

File: src/Compute/Compute.Test/ScenarioTests/VirtualMachineScaleSetTests.cs [hunk 5]
```diff
@@ -375,5 +375,13 @@ public void TestVirtualMachineScaleSetSecurityTypeUpdate()
             TestRunner.RunTestScript("Test-VirtualMachineScaleSetSecurityTypeUpdate");
         }
 
+        [Fact]
+        [Trait(Category.AcceptanceType, Category.CheckIn)]
```
[the remark this answers is not in the document]
Comment (haagha) [reply src/Compute/Compute.Test/ScenarioTests/VirtualMachineScaleSetTests.cs:379]: test has been fixed

File: src/Compute/Compute.Test/ScenarioTests/VirtualMachineScaleSetTests.ps1 [hunk 6]
```diff
@@ -4944,4 +4944,60 @@ function Test-VirtualMachineScaleSetSecurityTypeUpdate
...
+        $securePassword = ConvertTo-SecureString $adminPassword -AsPlainText -Force;
+        $cred = New-Object System.Management.Automation.PSCredential ($adminUsername, $securePassword);
+
+        $VmssFlex = New-AzVmss `
+          -ResourceGroupName $rgname `
+          -Name $vmssName `
+          -OrchestrationMode 'Flexible' `
+          -Location 'eastus' `
+          -Credential $cred `
+          -DomainNameLabel "scaleset-70f699"
```
Comment (Sandido) [inline src/Compute/Compute.Test/ScenarioTests/VirtualMachineScaleSetTests.ps1:4982]: better to not have this hardcoded, rather do something like d + $rgname.

File: src/Compute/Compute.Test/ScenarioTests/VirtualMachineTests.cs [hunk 7]
```diff
@@ -445,7 +445,7 @@ public void TestCapacityReservation()
             TestRunner.RunTestScript("Test-CapacityReservation");
         }
 
-        [Fact]
+        [Fact(Skip = "Cannot run on Github Platform due to ACL API")]
```
Comment (Sandido) [inline src/Compute/Compute.Test/ScenarioTests/VirtualMachineTests.cs:448]: wait, what happened here???

File: src/Compute/Compute/Compute.csproj [hunk 8]
```diff
@@ -15,6 +15,7 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.Azure.PowerShell.AutoMapper" Version="6.2.2" />
     <PackageReference Include="Microsoft.Azure.Management.ResourceGraph" Version="2.1.0" />
+    <PackageReference Include="System.IO.FileSystem.AccessControl" Version="5.0.0" />
```
Comment (Sandido) [inline src/Compute/Compute/Compute.csproj:18]: Why is this here?

File: src/Compute/Compute/VirtualMachine/Operation/NewAzureVMCommand.cs [hunk 9]
```diff
@@ -1456,6 +1458,23 @@ private string GenerateOrFindSshKey()
                     }
                     Console.WriteLine("Private key is saved to " + privateKeyFilePath);
 
+                    FileSecurity fileSecurity = new FileSecurity(privateKeyFilePath, AccessControlSections.Access);
```
Comment (Sandido) [inline src/Compute/Compute/VirtualMachine/Operation/NewAzureVMCommand.cs:1461]: Are these related to your PR??

**Review by Sandido — CHANGES_REQUESTED** [review]
some concerns with other changes being bundled in here by accident regarding SSHKey stuff.

File: src/Compute/Compute/VirtualMachine/Operation/NewAzureVMCommand.cs (hunk 9 above)
Comment (haagha) [reply src/Compute/Compute/VirtualMachine/Operation/NewAzureVMCommand.cs:1461]: yes, I'm merging changes for SSH key into this PR to avoid merge issues.

File: src/Compute/Compute/Compute.csproj (hunk 8 above)
Comment (haagha) [reply src/Compute/Compute/Compute.csproj:18]: Adding AccessControl for File Access Control to reduce permissions for private key

File: src/Compute/Compute.Test/ScenarioTests/VirtualMachineTests.cs (hunk 7 above)
Comment (haagha) [reply src/Compute/Compute.Test/ScenarioTests/VirtualMachineTests.cs:448]: Github tests are run platform independent so it does not support ACL API when generating ssh keys. Tests can be run fine locally.

File: src/Compute/Compute.Test/ScenarioTests/VirtualMachineScaleSetTests.ps1 (hunk 6 above)
Comment (haagha) [reply src/Compute/Compute.Test/ScenarioTests/VirtualMachineScaleSetTests.ps1:4982]: Does not affect the test in this case.

**Comment by haagha** [comment]
@Sandido SSH File permission change and Update-Azvm change need to be tested together, bundling these changes to avoid any merge issues.

**Review by Sandido — DISMISSED** [review]

**Comment by haagha** [comment]
@VeryEarly PR was approved, but I had to rerecord test because of @Sandido TL changes.

**Review by Sandido — APPROVED** [review]

**Review by VeryEarly — APPROVED** [review]
